### PR TITLE
Add Secure Enclaves dropdown with customizable Advanced options

### LINKED ISSUE #21869: Support 'Secure Enclaves' advanced options for MSSQL provider
## Problem Statement

When the user creates a new Connection in Azure Data Studio, they have the option to set Advanced Properties. The Always Encrypted Properties can also be configured under these Advanced Properties. However, the current properties are not clear and insufficient to be used with secure enclaves.  

The issues with the advanced properties are: 

- **Always Encrypted – Enabled or disabled**. This option doesn’t mention if the user wants to use Always Encrypted or Always Encrypted with secure enclaves. 

- **Attestation Protocol** – It is not clear for the user whether he/she needs to configure an attestation protocol related to the AE Enabled/Disabled setting. Secondly, the attestation protocol None is missing. This option should be added because you can use AE with secure enclaves in SQL Server 2022 without attestation. Currently AE with VBS enclaves in Azure SQL DB doesn’t support attestation so the protocol should always be None. 

- Always Encrypted related advanced options are not enabled/disabled based on user provided values.

## Proposed Solution

This improvement addresses the above problems and should be aligned with the Always Encrypted tab in the Connection Properties of SSMS experience.

### Add 'Secure Enclaves' dropdown

A new dropdown box “Secure Enclaves” should be added. This will make it clearer to the user to use Always Encrypted or Always Encrypted with Secure Enclaves. When the user opens the advanced properties, only the Always Encrypted dropdown box should be enabled. The other dropdown boxes that apply to Always Encrypted settings should be disabled.

When the user selects “Enabled” in the "Always Encrypted" dropdown, the dropdown box “Secure Enclaves” should become available. This will allow the user to configure Secure Enclaves.

When the user selects "Enabled" in the "Secure Enclaves" dropdown, the dropdown box "Attestation Protocol" should become available.

### Update 'Attestation Protocol' dropdown

This dropdown box should contain 4 values: 

- `<Blank>` (default)
- Host Guardian Service 
- Microsoft Azure Attestation 
- None 

Depending on the option that the user selects, the "Enclave Attestation URL" input box should be disabled or enabled. 

| Attestation Protocol | Enclave Attestation URL |
| ------ | -------- |
| `<Blank>` (default) | Disabled |
| Host Guardian Service | Enabled |
| Microsoft Azure Attestation | Enabled |
| None | Disabled |

If the user selects Host Guardian Service or Microsoft Azure Attestation, the Enclave Attestation URL cannot be empty. If the user doesn’t specify an URL when clicking the OK button, the following error message should be shown: 

_“Attestation URL cannot be empty with the selected value of Attestation Protocol.”_

### PR DESCRIPTION
**Pull request by cheenamalhotra**
This PR fixes #21869 along with PR https://github.com/microsoft/sqltoolsservice/pull/1880

+Introduces show/hide capabilities in Advanced Options.

With changes:

cc @jaszymas @Pietervanhove

### CODE REVIEW

File: src/sql/workbench/browser/modal/optionsDialog.ts [hunk 1]
```diff
@@ -204,13 +206,85 @@ export class OptionsDialog extends Modal {
...
+		for (let optionName in this._optionElements) {
+			let widget: Widget = this._optionElements[optionName].optionWidget;
+			if (widget instanceof SelectBox) {
+				this._registerSelectionChangeEvents([this._optionElements], this._optionElements[optionName].option, widget, container);
+			}
+		}
+	}
+
+	private _registerSelectionChangeEvents(collections: { [optionName: string]: OptionsDialogHelper.IOptionElement }[], option: azdata.ServiceOption, widget: SelectBox, container: HTMLElement) {
+		if (option && option.onSelectionChange) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/browser/modal/optionsDialog.ts:231]: nit
```suggestion
		if (option?.onSelectionChange) {
```

File: src/sql/workbench/browser/modal/optionsDialog.ts [hunk 2]
```diff
@@ -204,13 +206,85 @@ export class OptionsDialog extends Modal {
...
+	 * @returns Widget if found, undefined otherwise
+	 */
+	private _findWidget(collections: { [optionName: string]: OptionsDialogHelper.IOptionElement }[], id: string): AdsWidget | undefined {
+		let foundWidget: AdsWidget | undefined;
+		collections.forEach((collection) => {
+			if (!foundWidget) {
+				foundWidget = collection[id].optionWidget;
+			}
+		});
+		return foundWidget;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/browser/modal/optionsDialog.ts:285]:
```suggestion
		return collections.find(collection => collection[id].optionWidget);
```

**Review by Charles-Gagnon — APPROVED** [review]
Please also update https://github.com/microsoft/azuredatastudio/wiki/Contribution-points with this new functionality you're adding.

File: src/sql/workbench/browser/modal/optionsDialog.ts (hunk 2 above)
Comment (cheenamalhotra) [reply src/sql/workbench/browser/modal/optionsDialog.ts]: Oops, that didn't compile it seems..

File: src/sql/workbench/browser/modal/optionsDialog.ts [hunk 3]
```diff
@@ -204,13 +206,79 @@ export class OptionsDialog extends Modal {
...
+	}
+
+	/**
+	 * Finds Widget from provided collection of widgets using option name.
+	 * @param collections collections of widgets to search for the widget with the widget Id
+	 * @param id Widget Id
+	 * @returns Widget if found, undefined otherwise
+	 */
+	private _findWidget(collections: { [optionName: string]: OptionsDialogHelper.IOptionElement }[], id: string): AdsWidget | undefined {
+		return collections.find(collection => collection[id].optionWidget);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/browser/modal/optionsDialog.ts:279]: Oops, you'll need this instead
```suggestion
		return collections.find(collection => !!collection[id].optionWidget)?.optionWidget;
```

Comment (cheenamalhotra) [reply src/sql/workbench/browser/modal/optionsDialog.ts]: It needs another [id] lookup - I got it compiled, will check-in after testing

File: extensions/mssql/package.json [hunk 4]
```diff
@@ -919,15 +919,66 @@
...
+          "onSelectionChange": [
+            {
+              "values": [
+                "disabled",
+                ""
+              ],
+              "dependentOptionActions": [
+                {
+                  "optionName": "attestationProtocol",
+                  "action": "hide"
```
Comment (DBarmanMS) [inline extensions/mssql/package.json:977]: when the fields are being hid, do we clear the data in the those fields?

**Review by erinstellato-ms — COMMENTED** [review]
I cannot tell in the video if, when an option is disabled, the options/boxes below are also disabled/not editable.  I just want to make sure we do not allow users to select an option or try to type something, when it is not an appropriate based on a related option.

**Review by alanrenmsft — APPROVED** [review]

File: extensions/mssql/package.json (hunk 4 above)
Comment (cheenamalhotra) [reply extensions/mssql/package.json:977]: Yes, they're reset to default (empty) when hidden

**Comment by cheenamalhotra** [comment]
Hi Erin,

I've altered the experience to show/hide instead of disable/enable if that's confusing anything.. as the goal is to NOT allow users to modify options that don't apply.

**Comment by erinstellato-ms** [comment]
> Hi Erin,
> 
> I've altered the experience to show/hide instead of disable/enable if that's confusing anything.. as the goal is to NOT allow users to modify options that don't apply.

Ohhh, I do prefer the show/hide option.  Less chance for confusion :)

**Review by DBarmanMS — APPROVED** [review]

**Pull request merged**